src/main/java/filter/Parser.java

```diff
@@ -17,24 +17,62 @@
 
 public final class Parser {
 
-    private Parser(ArrayList<Token> input) {
+    private final List<Token> input;
+
+    /**
+     * Determines if the parser should fail gracefully if it finds the input
+     * in a incomplete state. See {@link #check}.
+     */
+    private final boolean isInCheckMode;
```

Comment:
Maybe `isCheckingSyntax`? 'Check mode' is defined purely by how the flag is used, so maybe we shouldn't try to make it a thing.